File: lib/services/lyrics_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models.dart';

/// Fetches lyrics from LRCLib (https://lrclib.net) — a free, CC0 lyrics
/// database that returns both synced (LRC) and plain text lyrics.
class LyricsService {
  static const _base = 'https://lrclib.net/api';

  Future<LyricResult> fetch({
    required String trackName,
    required String artistName,
    String? albumName,
    Duration? duration,
  }) async {
    // 1) Try the /get endpoint (requires all four params — most accurate).
    if (albumName != null && duration != null) {
      final r = await _get(Uri.parse(
        '$_base/get'
        '?track_name=${Uri.encodeQueryComponent(trackName)}'
        '&artist_name=${Uri.encodeQueryComponent(artistName)}'
        '&album_name=${Uri.encodeQueryComponent(albumName)}'
        '&duration=${duration.inSeconds}',
      ));
      if (r != null) return r;
    }
    // 2) Fall back to /search (partial match).
    final searchResp = await http.get(Uri.parse(
      '$_base/search'
      '?track_name=${Uri.encodeQueryComponent(trackName)}'
      '&artist_name=${Uri.encodeQueryComponent(artistName)}',
    ));
    if (searchResp.statusCode != 200) return LyricResult.empty;
    final list = jsonDecode(searchResp.body) as List;
    if (list.isEmpty) return LyricResult.empty;
    final first = list.first as Map<String, dynamic>;
    return _parse(first);
  }

  Future<LyricResult?> _get(Uri url) async {
    final resp = await http.get(url);
    if (resp.statusCode != 200) return null;
    final json = jsonDecode(resp.body) as Map<String, dynamic>;
    return _parse(json);
  }

  LyricResult _parse(Map<String, dynamic> json) {
    final synced = json['syncedLyrics'] as String?;
    final plain = json['plainLyrics'] as String? ?? '';
    if (synced != null && synced.isNotEmpty) {
      return LyricResult(
        synced: true,
        lines: _parseLrc(synced),
        plainText: plain,
      );
    }
    return LyricResult(synced: false, lines: const [], plainText: plain);
  }

  /// Parses an LRC-format string into timestamped lines.
  /// Only the timestamp/text structure is interpreted here; no lyric text
  /// is embedded in this source.
  static List<LyricLine> _parseLrc(String lrc) {
    final regex = RegExp(r'\[(\d+):(\d+)(?:\.(\d+))?\](.*)');
    final lines = <LyricLine>[];
    for (final raw in lrc.split('\n')) {
      final m = regex.firstMatch(raw);
      if (m == null) continue;
      final minutes = int.parse(m.group(1)!);
      final seconds = int.parse(m.group(2)!);
      final fracStr = m.group(3);
      int millis = 0;
      if (fracStr != null) {
        final f = fracStr.padRight(3, '0').substring(0, 3);
        millis = int.parse(f);
      }
      final text = (m.group(4) ?? '').trim();
      lines.add(LyricLine(
        Duration(minutes: minutes, seconds: seconds, milliseconds: millis),
        text,
      ));
    }
    lines.sort((a, b) => a.time.compareTo(b.time));
    return lines;
  }
}

File: Sources/Services/LyricsService.swift
//
//  LyricsService.swift
//  ViviMusic
//
//  LRCLib (https://lrclib.net) から歌詞を取得する。
//  認証不要・CC0 で、同期歌詞 (LRC) と通常テキストの両方が返る。
//

import Foundation

enum LyricsService {

    private static let base = "https://lrclib.net/api"

    private static let session: URLSession = {
        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = 15
        return URLSession(configuration: cfg)
    }()

    /// 曲に対応する歌詞を探す。
    /// まず厳密一致 (/get) を試し、無ければ部分一致 (/search) に落とす。
    static func fetch(for song: Song) async -> LyricResult {
        let started = Date()

        // 1) 厳密一致。アルバム名と長さが揃っている場合のみ使える。
        if let album = song.album, let duration = song.durationSeconds {
            if let result = await get(track: song.title,
                                      artist: song.artist,
                                      album: album,
                                      duration: duration) {
                EventLog.logDuration(.lyrics, videoID: song.id, start: started,
                                     message: "厳密一致で取得 (synced=\(result.synced))")
                return result
            }
        }

        // 2) 部分一致
        if let result = await search(track: song.title, artist: song.artist) {
            EventLog.logDuration(.lyrics, videoID: song.id, start: started,
                                 message: "検索一致で取得 (synced=\(result.synced))")
            return result
        }

        EventLog.logDuration(.lyrics, videoID: song.id, start: started,
                             message: "歌詞が見つかりませんでした")
        return .empty
    }

    // MARK: - API 呼び出し

    private static func get(track: String, artist: String,
                            album: String, duration: Int) async -> LyricResult? {
        var comps = URLComponents(string: "\(base)/get")
        comps?.queryItems = [
            URLQueryItem(name: "track_name", value: track),
            URLQueryItem(name: "artist_name", value: artist),
            URLQueryItem(name: "album_name", value: album),
            URLQueryItem(name: "duration", value: String(duration)),
        ]
        guard let url = comps?.url else { return nil }
        return await request(url: url, takeFirstOfArray: false)
    }

    private static func search(track: String, artist: String) async -> LyricResult? {
        var comps = URLComponents(string: "\(base)/search")
        comps?.queryItems = [
            URLQueryItem(name: "track_name", value: track),
            URLQueryItem(name: "artist_name", value: artist),
        ]
        guard let url = comps?.url else { return nil }
        return await request(url: url, takeFirstOfArray: true)
    }

    private static func request(url: URL, takeFirstOfArray: Bool) async -> LyricResult? {
        do {
            var req = URLRequest(url: url)
            req.setValue("ViviMusic/1.0 (iOS)", forHTTPHeaderField: "User-Agent")

            let (data, response) = try await session.data(for: req)
            let status = (response as? HTTPURLResponse)?.statusCode ?? -1
            guard status == 200 else {
                EventLog.log(.lyrics, message: "LRCLib HTTP \(status)")
                return nil
            }

            let json = JSON(data: data)
            let target = takeFirstOfArray ? json[0] : json
            guard target.exists else { return nil }
            return parse(target)

        } catch {
            EventLog.logError(.lyrics, error: error, context: "LRCLib 取得")
            return nil
        }
    }

    private static func parse(_ json: JSON) -> LyricResult? {
        let plain = json["plainLyrics"].string ?? ""
        if let synced = json["syncedLyrics"].string, !synced.isEmpty {
            return LyricResult(synced: true, lines: parseLRC(synced), plainText: plain)
        }
        if plain.isEmpty { return nil }
        return LyricResult(synced: false, lines: [], plainText: plain)
    }

    // MARK: - LRC パース

    /// `[mm:ss.xx] 歌詞` 形式を時刻つきの行に分解する。
    /// ここで解釈しているのはタイムスタンプの構造だけで、
    /// 歌詞テキストはソースコードには一切含まれない。
    static func parseLRC(_ lrc: String) -> [LyricLine] {
        let pattern = #"\[(\d+):(\d+)(?:\.(\d+))?\](.*)"#
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return [] }

        var lines: [LyricLine] = []
        for raw in lrc.components(separatedBy: .newlines) {
            let range = NSRange(raw.startIndex..., in: raw)
            guard let m = regex.firstMatch(in: raw, range: range) else { continue }

            func group(_ i: Int) -> String? {
                guard let r = Range(m.range(at: i), in: raw) else { return nil }
                return String(raw[r])
            }

            let minutes = Int(group(1) ?? "0") ?? 0
            let seconds = Int(group(2) ?? "0") ?? 0
            // 小数部は桁数がまちまちなので 3 桁に正規化する
            var millis = 0
            if let frac = group(3) {
                let padded = frac.padding(toLength: 3, withPad: "0", startingAt: 0)
                millis = Int(padded) ?? 0
            }
            let text = (group(4) ?? "").trimmingCharacters(in: .whitespaces)

            let time = TimeInterval(minutes * 60 + seconds) + TimeInterval(millis) / 1000
            lines.append(LyricLine(time: time, text: text))
        }
        return lines.sorted { $0.time < $1.time }
    }
}
